duckduckgo/Android
# Remove cookies preserving fireproof websites

Task/Issue URL: https://app.asana.com/0/0/1170688450957740/f
Tech Design URL: https://app.asana.com/0/481882893211075/1170688450957739/f
CC: 

## **Description**:

**What's included in this PR?**
Implement logic to remove cookies preserving the ones related to fireproof website.

When clearing cookies:

- Try to directly remove cookies from WebView database, preserving the ones with hosts related to a fireproof website
- If process fails, fallback to remove all the cookies to avoid any leak
- Send pixels in the following scenarios:
- database path not found
- database can't be opened
- delete query fails

**What's NOT included in this PR?**
- UI improvements from design feedback
- Pixels from user interactions with Fireproof options

## **Steps to test this PR**:

#### Test 1: No cookies preserved
1. Fresh install
1. Visit some different websites
1. Press fire button
1. Access device file explorer
1. Ensure cookie database doesn't exist or database is empty

#### Test 2: DDG preserved
1. Fresh install
1. Visit duckduckgo website
1. Set a theme (it will create a cookies)
1. Press fire button
1. Visit duckduckgo again
1. Ensure theme has been preserved

#### Test 3: Fireproof website preserved
1. Fresh install
1. Visit any website (e.g: twitter.com)
1. Perform an action that will create cookies (e.g: login)
1. Fireproof website
1. Press fire button
1. Visit fireproof website
1. Ensure state has been preserved (e.g: you are logged in)

#### Test 4: Database not found: Database not found, all cookies removed
1. Replace following line from `WebViewDatabaseLocator`:
Line 28: `private val knownLocations = listOf("/fakePath")`
1. Fresh install
1. Visit any website (e.g: twitter.com)
1. Press fire button
1. Ensure pixel `m_cdb_nf` sent
1. Ensure captured and recorded
1. Access device file explorer
1. Ensure cookie database doesn't exist or database is empty

#### Test 5: Error opening database, all cookies removed
1. Replace the following line from `SQLCookieRemover`, line 70:
```
            SQLiteDatabase.openDatabase(
                "fakepath",
                null,
                SQLiteDatabase.OPEN_READWRITE,
                DatabaseErrorHandler { Timber.e("COOKIE: onCorruption") })
```
1. Fresh install
1. Visit any website (e.g: twitter.com)
1. Press fire button
1. Ensure pixel `m_cdb_oe` sent
1. Ensure captured and recorded
1. Access device file explorer
1. Ensure cookie database doesn't exist or database is empty

#### Test 6: Error deleting cookies, all cookies removed
1. Replace the following line from `SQLCookieRemover`, line 87:
```
val number = delete("faketable", whereClause, excludedSites.toTypedArray())
```
1. Fresh install
1. Visit any website (e.g: twitter.com)
1. Press fire button
1. Ensure pixel `m_cdb_de` sent
1. Ensure captured and recorded
1. Access device file explorer
1. Ensure cookie database doesn't exist or database is empty

Additional tests:
Instead of using firebutton, you can use auto cleaner data from settings.

---
###### Internal references:
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)
